rasa/nlu/test.py

```diff
@@ -442,8 +442,73 @@ def substitute_labels(labels, old, new):
     return [new if label == old else label for label in labels]
 
 
+def collect_ner_results(utterance_targets, utterance_predictions, ner_filename):
+    """Log messages which result in both successful and
+       unsuccessful entity predictions and save them to file"""
+
+    # there should be a finite number of utterances
+    if (
+        utterance_targets is None
+        or utterance_predictions is None
+        or len(utterance_targets) != len(utterance_predictions)
+    ):
+        logger.debug("Utterance mismatch.  Cannot report on entities")
+        return
+
+    # list of pairs.  The first is the expected.  The second is predicted.
+    # None is present if one of the two is not relevant
+    tps = []
+    fps = []
+    fns = []
+
+    logger.info("Iterating through utterances for entity mentions")
+    for te, pe in zip(utterance_targets, utterance_predictions):
+        target_count = 0
+        predicted_count = 0
+        while target_count < len(te) or predicted_count < len(pe):
+            target_entity = None if target_count == len(te) else te[target_count]
+            predicted_entity = (
+                None if predicted_count == len(pe) else pe[predicted_count]
+            )
+            if predicted_entity is None or (
+                target_entity is not None
+                and target_entity["start"] < predicted_entity["start"]
+            ):
+                fns.append((target_entity, None))
+                target_count += 1
+            elif target_entity is None or (
+                predicted_entity is not None
+                and predicted_entity["start"] < target_entity["start"]
+            ):
+                fps.append((None, predicted_entity))
+                predicted_count += 1
+            else:
+                target_type = target_entity["entity"]
+                predicted_type = predicted_entity["entity"]
+                if (
+                    predicted_type == target_type
+                    and predicted_entity["end"] == target_entity["end"]
+                ):
+                    tps.append((target_entity, predicted_entity))
+                else:
+                    fps.append((target_entity, predicted_entity))
+                    fns.append((target_entity, predicted_entity))
+                target_count += 1
+                predicted_count += 1
+
+    logger.info("Writing entity mentions to file")
+    ner_dict = {"TP": [], "FP": [], "FN": []}
+    for tp in tps:
+        ner_dict["TP"].append({"Expected": tp[0], "Predicted": tp[1]})
```

Comment:
Can we just add to the dict inside of the logic instead of creating lists and then adding the lists to the dict?